lib/tools/adb-commands.js

```diff
@@ -97,13 +98,33 @@ methods.grantAllPermissions = async function (pkg, apk) {
     const requestedPermissions = await this.getReqPermissions(pkg, stdout);
     const grantedPermissions = await this.getGrantedPermissions(pkg, stdout);
     const permissonsToGrant = requestedPermissions.filter((x) => grantedPermissions.indexOf(x) < 0);
-    let cmd = [];
+    if (!permissonsToGrant.length) {
+      log.info(`${pkg} contains no permissions available for granting.`);
+      return true;
+    }
+    // As it consumes more time for granting each permission,
+    // trying to grant all permission by forming equivalent command.
+    // Also, it is necessary to split long commands into chunks, since the maximum length of
+    // adb shell buffer is limited
+    let cmds = [];
+    let cmdChunk = [];
     for (let permission of permissonsToGrant) {
-      //As it consumes more time for granting each permission,
-      //trying to grant all permission by forming equivalent command.
-      cmd.push('pm', 'grant', pkg, permission, ';');
+      const nextCmd = ['pm', 'grant', pkg, permission, ';'];
+      if (nextCmd.join(' ').length + cmdChunk.join(' ').length >= MAX_SHELL_BUFFER_LENGTH) {
+        cmds.push(cmdChunk);
+        cmdChunk = [];
+      }
+      cmdChunk = cmdChunk.concat(nextCmd);
+    }
+    if (cmdChunk.length) {
+      cmds.push(cmdChunk);
+    }
+    log.debug(`Got the following command chunks to execute: ${cmds}`);
+    let result = true;
+    for (let cmd of cmds) {
+      result = await this.shell(cmd) && result;
```

Comment:
@mykola-mokhnach would it be good to know which permission failed, what do you think ?